openstudiocore/src/utilities/filetypes/EpwFile.cpp

```diff
@@ -2202,204 +2866,1371 @@ namespace openstudio{
+        return boost::none;
+      }
+      ifs.close();
+    }
+    EpwComputedField id;
+    try {
+      id = EpwComputedField(name);
+    }
+    catch (...) {
+      LOG(Warn, "Unrecognized computed data field '" << name << "'");
+      return boost::none;
+    }
+
+    std::string units = EpwDataPoint::getUnits(id);
+    boost::optional<double>(EpwDataPoint::*compute)() const;
+    switch (id.value()) {
+      case EpwComputedField::SaturationPressure:
+        compute = &EpwDataPoint::saturationPressure;
+        break;
+      case EpwComputedField::Enthalpy:
+        compute = &EpwDataPoint::enthalpy;
+        break;
+      case EpwComputedField::HumidityRatio:
+        compute = &EpwDataPoint::humidityRatio;
+        break;
+      case EpwComputedField::WetBulbTemperature:
+        compute = &EpwDataPoint::wetbulb;
+        break;
+      case EpwComputedField::Density:
+        compute = &EpwDataPoint::density;
+        break;
+      case EpwComputedField::SpecificVolume:
+        compute = &EpwDataPoint::specificVolume;
+        break;
+      default:
+        return boost::none;
+    }
+    DateTimeVector dates;
+    dates.push_back(DateTime()); // Use a placeholder to avoid an insert
+    std::vector<double> values;
+    for (unsigned int i = 0; i<m_data.size(); i++) {
+      Date date = m_data[i].date();
+      Time time = m_data[i].time();
+      boost::optional<double> value = (m_data[i].*compute)();
+      if (value) {
+        dates.push_back(DateTime(date, time));
+        values.push_back(value.get());
+      }
+    }
+    if (values.size()) {
+      DateTime start = dates[1] - Time(0, 0, 0, 3600.0 / m_recordsPerHour);
+      dates[0] = start; // Overwrite the placeholder
+      return boost::optional<TimeSeries>(TimeSeries(dates, openstudio::createVector(values), units));
+    }
+    return boost::none;
+  }
+
+  bool EpwFile::translateToWth(openstudio::path path, std::string description)
+  {
+    if(m_data.size()==0) {
+      if (!openstudio::filesystem::exists(m_path) || !openstudio::filesystem::is_regular_file(m_path)){
+        LOG_AND_THROW("Path '" << m_path << "' is not an EPW file");
+      }
+
+      // set checksum
+      m_checksum = openstudio::checksum(m_path);
+
+      // open file
+      std::ifstream ifs(openstudio::toString(m_path));
+
+      if (!parse(ifs, true)) {
+        ifs.close();
+        LOG(Error,"EpwFile '" << toString(m_path) << "' cannot be processed");
+        return false;
+      }
+      ifs.close();
+    }
+
+    if(description.empty()) {
+      description = "Translated from " + openstudio::toString(this->path());
+    }
+
+    if(!data().size()) {
+      LOG(Error, "EPW file contains no data to translate");
+      return false;
+    }
+
+    openstudio::filesystem::ofstream fp(path, std::ios_base::binary);
+    if(!fp.is_open()) {
+      LOG(Error, "Failed to open file '" + openstudio::toString(path) + "'");
+      return false;
+    }
+
+    fp << "WeatherFile ContamW 2.0\n";
```

Comment:
Is this from @joseph-robertson or @jasondegraw ?

Reply:
That's from me... some years ago.